yjit/src/codegen.rs

```diff
@@ -3599,6 +3599,43 @@ fn jit_rb_obj_equal(
     true
 }
 
+/// If string is frozen, duplicate it to get a non-frozen string. Otherwise, return it.
+fn jit_rb_str_uplus(
+    _jit: &mut JITState,
+    ctx: &mut Context,
+    cb: &mut CodeBlock,
+    _ocb: &mut OutlinedCb,
+    _ci: *const rb_callinfo,
+    _cme: *const rb_callable_method_entry_t,
+    _block: Option<IseqPtr>,
+    _argc: i32,
+    _known_recv_class: *const VALUE,
+) -> bool
+{
+    let recv = ctx.stack_pop(1);
+
+    add_comment(cb, "Unary plus on string");
+    mov(cb, REG0, recv);
+    mov(cb, REG1, mem_opnd(64, REG0, RUBY_OFFSET_RBASIC_FLAGS));
+    test(cb, REG1, imm_opnd(RUBY_FL_FREEZE as i64));
+
+    let ret_label = cb.new_label("stack_ret".to_string());
+    // If the string isn't frozen, we just return it. It's already in REG0.
+    jz_label(cb, ret_label);
+
+    // Str is frozen - duplicate
+    mov(cb, C_ARG_REGS[0], REG0);
+    call_ptr(cb, REG0, rb_str_dup as *const u8);
+    // Return value is in REG0, drop through and return it.
+
+    cb.write_label(ret_label);
+    let stack_ret = ctx.stack_push(Type::String);
+    mov(cb, stack_ret, REG0);
+
+    cb.link_labels();
+    true
```

Comment:
Hi Noah. I think it's a good initiative to add support for unary plus 👍 

The one thing I will say is that in general I try to avoid having conditional branches in the generated code. It's much better if the code just falls through without jumping forwards or backwards (except for outlined code aka guards). My main worry is that CPU branch predictors are actually a finite resource. I can explain in more detail on slack if you want.

In this case I think just calling `str_uplus(VALUE str)` would be more compact and add less branches to the code, but I will merge the PR regadless because this is probably fairly infrequent and so the impact on code size will be very small.

Reply:
That makes sense. But when str_uplus is called, it immediately does an if/else on the same condition. Does that have a different effect on branch predictors? I don't know if doing a function call somehow resets the limit...